codegen/src/main/twirl/templates/ScalaServer/Handler.scala.txt

```diff
@@ -2,86 +2,94 @@
-    partial(implementation, @{service.name}.name)
+    /**
+     * Creates a partial `HttpRequest` to `HttpResponse` handler that can be combined with handlers of other
+     * services with `akka.grpc.scaladsl.ServiceHandler.concatOrNotFound` and then used in for example
+     * `Http().bindAndHandleAsync`.
+     *
+     * Use `@{serviceName}Handler.apply` if the server is only handling one service.
+     */
+    def partial(implementation: @serviceName)(implicit mat: Materializer): PartialFunction[HttpRequest, Future[HttpResponse]] =
+      partial(implementation, @{service.name}.name)
 
-  /**
-   * Creates a partial `HttpRequest` to `HttpResponse` handler that can be combined with handlers of other
-   * services with `akka.grpc.scaladsl.ServiceHandler.concatOrNotFound` and then used in for example
-   * `Http().bindAndHandleAsync`.
-   *
-   * Use `@{service.name}Handler.apply` if the server is only handling one service.
-   *
-   * Registering a gRPC service under a custom prefix is not widely supported and strongly discouraged by the specification.
-   */
-  def partial(implementation: @service.name, prefix: String)(implicit mat: Materializer): PartialFunction[HttpRequest, Future[HttpResponse]] = {
-    implicit val ec: ExecutionContext = mat.executionContext
-    import @{service.name}.Serializers._
+    /**
+     * Creates a partial `HttpRequest` to `HttpResponse` handler that can be combined with handlers of other
+     * services with `akka.grpc.scaladsl.ServiceHandler.concatOrNotFound` and then used in for example
+     * `Http().bindAndHandleAsync`.
+     *
+     * Use `@{serviceName}Handler.apply` if the server is only handling one service.
+     *
+     * Registering a gRPC service under a custom prefix is not widely supported and strongly discouraged by the specification.
+     */
+    def partial(implementation: @serviceName, prefix: String)(implicit mat: Materializer): PartialFunction[HttpRequest, Future[HttpResponse]] = {
+      implicit val ec: ExecutionContext = mat.executionContext
+      import @{service.name}.Serializers._
 
-    def handle(request: HttpRequest, method: String): Future[HttpResponse] = method match {
-      @for(method <- service.methods) {
-      case "@method.grpcName" =>
-        val responseCodec = Codecs.negotiate(request)
-        @{method.unmarshal}(request)(@method.deserializer.name, mat)
-          .@{if(method.outputStreaming) { "map" } else { "flatMap" }}(implementation.@{method.name}(_))
-          .map(e => @{method.marshal}(e)(@method.serializer.name, mat, responseCodec))
+      def handle(request: HttpRequest, method: String): Future[HttpResponse] = method match {
+        @for(method <- service.methods) {
+        case "@method.grpcName" =>
+          val responseCodec = Codecs.negotiate(request)
+          @{if(powerApis) { "val metadataMap = MetadataMap(request.headers)" } else { "" }}
```

Comment:
Now this is eager for all power APIs even if they don't actually read the metadata. An actual type inbetween could lessen that to just one object allocation, and then parse the headers into our datatypes only if they are accessed.

Reply:
Reverted to using an enhanced version of the `Metadata` trait, see `runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala`.